File: android/src/main/kotlin/com/superwall/superwallkit_flutter/SuperwallkitFlutterPlugin.kt
package com.superwall.superwallkit_flutter

import android.app.Activity
import android.os.Debug
import android.util.Log
import com.superwall.sdk.misc.runOnUiThread
import com.superwall.superwallkit_flutter.bridges.BridgeId
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.suspendCancellableCoroutine
import java.util.WeakHashMap
import java.util.concurrent.atomic.AtomicInteger
import java.util.concurrent.atomic.AtomicReference
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException

// Application lifecycle tracking enum
enum class ApplicationLifecycle {
    FOREGROUND, BACKGROUND, TRANSITIONING
}

class SuperwallkitFlutterPlugin : FlutterPlugin, ActivityAware {
    var currentActivity: Activity? = null
    
    // Add application lifecycle tracking
    private val applicationLifecycle = AtomicReference(ApplicationLifecycle.BACKGROUND)

    companion object {
        var instance: SuperwallkitFlutterPlugin? = null
            private set
        val reattachementCount = AtomicInteger(0)
        val lock = Object()
        val currentActivity: Activity?
            get() = instance?.currentActivity
    }

    init {
        if (BuildConfig.DEBUG && BuildConfig.WAIT_FOR_DEBUGGER) {
            Debug.waitForDebugger()
        }

        // Only allow instance to get set once.
        if (instance == null) {
            instance = this
        }
    }

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        synchronized(lock) {
            BridgingCreator.setFlutterPlugin(flutterPluginBinding)
            reattachementCount.incrementAndGet()
            println("SuperwallkitFlutterPlugin: onAttachedToEngine - reattachment count: ${reattachementCount.get()}")
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        applicationLifecycle.set(ApplicationLifecycle.TRANSITIONING)
        CoroutineScope(Dispatchers.Main).launch {
            try {
                // Call tearDown, but our modified version preserves the instance
                BridgingCreator.shared().tearDown()
                println("SuperwallkitFlutterPlugin: onDetachedFromEngine - plugin instance preserved")
                // Don't set instance to null - we need to maintain the reference
            } catch (e: Exception) {
                println("Error during plugin detachment: ${e.message}")
                e.printStackTrace()
            }
        }
    }

    //region ActivityAware

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        currentActivity = binding.activity
        applicationLifecycle.set(ApplicationLifecycle.FOREGROUND)
        
        // Proactively check and reattach if needed when coming to foreground
        CoroutineScope(Dispatchers.Main).launch {
            try {
                if (!BridgingCreator.checkPluginHealth()) {
                    // App is coming to foreground, ensure plugin is registered
                    println("SuperwallkitFlutterPlugin: Plugin health check failed, forcing reattachment")
                    val flutterBinding = BridgingCreator.lastKnownBinding
                    BridgingCreator.forceReattachment(flutterBinding)
                } else {
                    println("SuperwallkitFlutterPlugin: Plugin health check passed")
                }
            } catch (e: Exception) {
                println("Error during health check in onAttachedToActivity: ${e.message}")
                e.printStackTrace()
            }
        }
    }

    override fun onDetachedFromActivityForConfigChanges() {
        applicationLifecycle.set(ApplicationLifecycle.TRANSITIONING)
        currentActivity = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        currentActivity = binding.activity
        applicationLifecycle.set(ApplicationLifecycle.FOREGROUND)
        
        // Check health after configuration changes too
        CoroutineScope(Dispatchers.Main).launch {
            try {
                if (!BridgingCreator.checkPluginHealth()) {
                    println("SuperwallkitFlutterPlugin: Plugin health check failed after config change, forcing reattachment")
                    val flutterBinding = BridgingCreator.lastKnownBinding
                    BridgingCreator.forceReattachment(flutterBinding)
                }
            } catch (e: Exception) {
                println("Error during health check in onReattachedToActivityForConfigChanges: ${e.message}")
                e.printStackTrace()
            }
        }
    }

    override fun onDetachedFromActivity() {
        applicationLifecycle.set(ApplicationLifecycle.BACKGROUND)
        currentActivity = null
    }

    //endregion
}

fun <T> MethodCall.argumentForKey(key: String): T? {
    return this.argument(key)
}

// Enhanced bridgeInstance with better error handling
suspend fun <T> MethodCall.bridgeInstance(key: String): T? {
    BreadCrumbs.append("SuperwallKitFlutterPlugin.kt: Invoke bridgeInstance(key:) on $this. Key is $key")
    val bridgeId = this.argument<String>(key)
    if (bridgeId == null) {
        Log.e("SWKP", "No bridgeId found for key: $key")
        return null
    }
    
    try {
        BreadCrumbs.append("SuperwallKitFlutterPlugin.kt: Invoke bridgeInstance(key:) in on $this. Found bridgeId $bridgeId")
        return BridgingCreator.shared().bridgeInstance(bridgeId)
    } catch (e: Exception) {
        // If we can't get the instance, try to force reattachment and retry once
        Log.e("SWKP", "Error getting bridge instance: ${e.message}")
        BridgingCreator.forceReattachment(BridgingCreator.lastKnownBinding)
        kotlinx.coroutines.delay(300) // Brief delay to allow reattachment
        
        // One more try
        try {
            return BridgingCreator.shared().bridgeInstance(bridgeId)
        } catch (e2: Exception) {
            Log.e("SWKP", "Failed to get bridge instance after reattachment: ${e2.message}")
            throw e2
        }
    }
}

suspend fun <T> BridgeId.bridgeInstance(): T? {
    BreadCrumbs.append("SuperwallKitFlutterPlugin.kt: Invoke bridgeInstance() in on $this")
    try {
        return BridgingCreator.shared().bridgeInstance(this)
    } catch (e: Exception) {
        // If we can't get the instance, try to force reattachment and retry once
        Log.e("SWKP", "Error getting bridge instance: ${e.message}")
        BridgingCreator.forceReattachment(BridgingCreator.lastKnownBinding)
        kotlinx.coroutines.delay(300) // Brief delay to allow reattachment
        
        // One more try
        return BridgingCreator.shared().bridgeInstance(this)
    }
}

fun MethodChannel.Result.badArgs(call: MethodCall) {
    return badArgs(call.method)
}

fun MethodChannel.Result.badArgs(method: String) {
    return error("BAD_ARGS", "Missing or invalid arguments for '$method'", null)
}

fun MethodChannel.invokeMethodOnMain(method: String, arguments: Any? = null) {
    runOnUiThread {
        try {
            invokeMethod(method, arguments);
        } catch (e: Exception) {
            Log.e("SWKP", "Error invoking method $method: ${e.message}")
        }
    }
}

suspend fun MethodChannel.asyncInvokeMethodOnMain(method: String, arguments: Any? = null): Any? =
    suspendCancellableCoroutine { continuation ->
        runOnUiThread {
            try {
                invokeMethod(method, arguments, object : MethodChannel.Result {
                    override fun success(result: Any?) {
                        if (!continuation.isCompleted) {
                            continuation.resume(result)
                        }
                    }

                    override fun error(errorCode: String, errorMessage: String?, errorDetails: Any?) {
                        if (!continuation.isCompleted) {
                            continuation.resumeWithException(
                                RuntimeException("Error invoking method: $errorCode, $errorMessage")
                            )
                        }
                    }

                    override fun notImplemented() {
                        if (!continuation.isCompleted) {
                            continuation.resumeWithException(
                                UnsupportedOperationException("Method not implemented: $method")
                            )
                        }
                    }
                })
            } catch (e: Exception) {
                if (!continuation.isCompleted) {
                    continuation.resumeWithException(e)
                }
            }
        }
    }

fun <T> Map<String, Any?>.argument(key: String): T? {
    return this[key] as? T
}

object MethodChannelAssociations {
    private val bridgeIds = WeakHashMap<MethodChannel, String>()

    fun setBridgeId(methodChannel: MethodChannel, bridgeId: String) {
        bridgeIds[methodChannel] = bridgeId
    }

    fun getBridgeId(methodChannel: MethodChannel): String {
        return bridgeIds[methodChannel]
            ?: throw IllegalStateException("bridgeId must be set at initialization of MethodChannel")
    }
}

fun MethodChannel.setBridgeId(bridgeId: String) {
    MethodChannelAssociations.setBridgeId(this, bridgeId)
}

fun MethodChannel.getBridgeId(): String {
    return MethodChannelAssociations.getBridgeId(this)
}

File: ios/Classes/SuperwallkitFlutterPlugin.swift
import Flutter
import UIKit
import SuperwallKit

// Application lifecycle tracking enum
enum ApplicationLifecycle {
  case foreground
  case background
  case transitioning
}

public class SuperwallkitFlutterPlugin: NSObject, FlutterPlugin {
  private static var alreadyRegistered = false
  
  // Track application lifecycle
  private static var applicationLifecycle: ApplicationLifecycle = .background
  
  // Count for debugging purposes
  private static var reattachmentCount = 0

  public static func register(with registrar: FlutterPluginRegistrar) {
    // This should get called on the main thread by default
    if alreadyRegistered {
      // Even if already registered, update the registrar for potential reattachment
      reattachmentCount += 1
      print("SuperwallkitFlutterPlugin: register called again, count: \(reattachmentCount)")
      return
    }
    
    alreadyRegistered = true
    print("SuperwallkitFlutterPlugin: First registration")

    // Register for app lifecycle notifications
    NotificationCenter.default.addObserver(
      forName: UIApplication.didEnterBackgroundNotification, 
      object: nil, 
      queue: nil
    ) { _ in
      applicationLifecycle = .background
      print("SuperwallkitFlutterPlugin: App entered background")
    }
    
    NotificationCenter.default.addObserver(
      forName: UIApplication.willEnterForegroundNotification, 
      object: nil, 
      queue: nil
    ) { _ in
      applicationLifecycle = .transitioning
      print("SuperwallkitFlutterPlugin: App will enter foreground")
      
      // Check plugin health when coming to foreground
      DispatchQueue.main.async {
        if !BridgingCreator.checkPluginHealth() {
          print("SuperwallkitFlutterPlugin: Plugin health check failed, forcing reattachment")
          _ = BridgingCreator.forceReattachment()
        } else {
          print("SuperwallkitFlutterPlugin: Plugin health check passed")
        }
      }
    }
    
    NotificationCenter.default.addObserver(
      forName: UIApplication.didBecomeActiveNotification, 
      object: nil, 
      queue: nil
    ) { _ in
      applicationLifecycle = .foreground
      print("SuperwallkitFlutterPlugin: App became active")
    }

    BridgingCreator.register(with: registrar)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    result(FlutterMethodNotImplemented)
  }
  
  deinit {
    NotificationCenter.default.removeObserver(self)
  }
}

extension FlutterMethodCall {
  func argument<T>(for key: String) -> T? {
    return (arguments as? [String: Any])?[key] as? T
  }

  // Enhanced bridgeInstance with retry
  func bridgeInstance(for key: String) -> Any? {
    guard let bridgeId: String = argument(for: key) else {
      return nil
    }

    // First attempt
    if let instance = BridgingCreator.shared.bridgeInstance(for: bridgeId) {
      return instance
    }
    
    // Not found, try reattachment and try again
    if BridgingCreator.forceReattachment() {
      print("Reattached plugin, trying to get bridge instance again")
      if let instance = BridgingCreator.shared.bridgeInstance(for: bridgeId) {
        return instance
      }
    }
    
    return nil
  }
  
  // Type-specific convenience method
  func bridgeInstanceTyped<T>(for key: String) -> T? {
    guard let instance = bridgeInstance(for: key) else {
      return nil
    }
    return instance as? T
  }
}

extension BridgeId {
  func bridgeInstance() -> Any? {
    // First attempt
    if let instance = BridgingCreator.shared.bridgeInstance(for: self) {
      return instance
    }
    
    // Not found, try reattachment and try again
    if BridgingCreator.forceReattachment() {
      print("Reattached plugin, trying to get bridge instance again")
      if let instance = BridgingCreator.shared.bridgeInstance(for: self) {
        return instance
      }
    }
    
    return nil
  }
  
  // Type-specific convenience method
  func bridgeInstanceTyped<T>() -> T? {
    guard let instance = bridgeInstance() else {
      return nil
    }
    return instance as? T
  }
}

extension FlutterMethodCall {
  var badArgs: FlutterError {
    return FlutterError(code: "BAD_ARGS", message: "Missing or invalid arguments for '\(method)'", details: nil)
  }
}

extension FlutterMethodChannel {
  func invokeMethodOnMain(_ method: String, arguments: Any? = nil) {
    Task { @MainActor [weak self] in
      guard let self else { return }
      do {
        invokeMethod(method, arguments: arguments)
      } catch {
        print("Error invoking method \(method): \(error)")
      }
    }
  }

  @MainActor
  @discardableResult func asyncInvokeMethodOnMain(_ method: String, arguments: Any? = nil) async -> Any? {
    return await withCheckedContinuation { [weak self] continuation in
      guard let self else { 
        continuation.resume(returning: nil)
        return
      }
      
      do {
        invokeMethod(method, arguments: arguments) { result in
          continuation.resume(returning: result)
        }
      } catch {
        continuation.resume(returning: nil)
      }
    }
  }
}

extension Dictionary where Key == String {
  func argument<T>(for key: String) -> T? {
    return self[key] as? T
  }
}

extension FlutterMethodChannel {
  private static var bridgeIdKey: UInt8 = 0

  var bridgeId: String {
    get {
      guard let name = objc_getAssociatedObject(self, &FlutterMethodChannel.bridgeIdKey) as? String else {
        assertionFailure("bridgeId must be set at initialization of FlutterMethodChannel")
        return ""
      }

      return name
    }
    set(newValue) {
      objc_setAssociatedObject(self, &FlutterMethodChannel.bridgeIdKey, newValue, objc_AssociationPolicy.OBJC_ASSOCIATION_RETAIN_NONATOMIC)
    }
  }
}
